### PR TITLE
[SPARK-43290][SQL] Adds AES IV and AAD support to ExpressionImplUtils

### PR DESCRIPTION
**Pull request by sweisdb**
### What changes were proposed in this pull request?
This change adds support for optional IV and AAD fields to ExpressionImplUtils, which is the underlying library to support `aes_encrypt` and `aes_decrypt`. This allows callers to specify their own initialization vector values for some specific use cases, and to take advantage of AES-GCM's authenticated additional data optional input.

This change does **not** add the support to the user-facing `aes_encrypt` and `aes_decrypt` yet. That will be added in a follow-up, rather than in a single complex change.

### Why are the changes needed?

There are some use cases where callers to ExpressionImplUtils via aes_encrypt may want to provide initialization vectors (IVs) or additional authenticated data (AAD). The most common cases will be:
1. Ensuring that ciphertext matches values that have been encrypted by external tools. In those cases, the caller will need to provide an identical IV value.
2. For AES-CBC mode, there are some cases where callers want to generate deterministic encrypted output.
3. For AES-GCM mode, providing AAD fields allows callers to bind additional data to an encrypted ciphertext so that it can only be decrypted by a caller providing the same value. This is often used to enforce some context.

### Does this PR introduce _any_ user-facing change?

Not yet. This change adds support to the underlying implementation, but does not yet update the SQL support to include the new parameters.

### How was this patch tested?

All existing unit tests still pass and new tests in `ExpressionImplUtilsSuite` exercise the new code paths:
```
build/sbt "sql/test:testOnly org.apache.spark.sql.DataFrameFunctionsSuite"
build/sbt "catalyst/test:testOnly org.apache.spark.sql.catalyst.expressions.ExpressionImplUtilsSuite"
```

### CODE REVIEW

File: sql/core/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionImplUtilsSuite.scala [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+  private val b64decoder = java.util.Base64.getDecoder
+  private val b64encoder = java.util.Base64.getEncoder
+
+  case class TestCase(
+    plaintext: String,
+    key: String,
+    base64CiphertextExpected: String,
+    mode: String,
+    padding: String = "Default",
+    ivHex: String = null,
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionImplUtilsSuite.scala:33]: nit: can't you follow Scala style and avoid `null`s? Like:
```scala
    ivHex: Option[String] = None,
...
    val ivBytes = ivHex.map(iv => Hex.unhex(iv.getBytes("UTF-8")))
```

File: sql/core/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionImplUtilsSuite.scala [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+  )
+
+  test("AesEncrypt Expected Errors") {
+    corruptedCiphertexts.foreach { t =>
+      val e1 = intercept[SparkRuntimeException] {
+        decOnlyCase(t)
+      }
+      assert(e1.isInstanceOf[SparkRuntimeException])
+      assert(e1.getErrorClass == t.expectedErrorClass)
+    }
```
Comment (MaxGekk) [inline sql/core/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionImplUtilsSuite.scala:258]: Could you use `checkError()` like:
```scala
      checkError(
        exception = intercept[SparkRuntimeException] {
          decOnlyCase(t)
        },
        errorClass = t.expectedErrorClass)
```

File: sql/core/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionImplUtilsSuite.scala (hunk 1 above)
Comment (sweisdb) [reply sql/core/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionImplUtilsSuite.scala]: Done

File: sql/core/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionImplUtilsSuite.scala (hunk 2 above)
Comment (sweisdb) [reply sql/core/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionImplUtilsSuite.scala]: Done

**Review by MaxGekk — CHANGES_REQUESTED** [review]
@sweisdb Could you fix the build errors:
```
/home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionImplUtilsSuite.scala:41:67: not found: value ivHex
[error]     val deterministic: Boolean = mode.equalsIgnoreCase("ECB") || (ivHex != null)
[error]                                                                   ^
[error] 
```

**Comment by sweisdb** [comment]
> @sweisdb Could you fix the build errors:
> 
> ```
> /home/runner/work/spark/spark/sql/core/src/test/scala/org/apache/spark/sql/catalyst/expressions/ExpressionImplUtilsSuite.scala:41:67: not found: value ivHex
> [error]     val deterministic: Boolean = mode.equalsIgnoreCase("ECB") || (ivHex != null)
> [error]                                                                   ^
> [error] 
> ```

Oops, sorry. I think I merged this in and didn't re-run tests after. I pushed a fix.

**Comment by sweisdb** [comment]
I rebased on master, which moves the location of the test suite. I ended up squashing commits by mistake and force pushed a single commit.

`build/sbt "catalyst/test:testOnly org.apache.spark.sql.catalyst.expressions.ExpressionImplUtilsSuite"`

File: core/src/main/resources/error/error-classes.json [hunk 3]
```diff
@@ -1051,6 +1051,16 @@
           "expects a binary value with 16, 24 or 32 bytes, but got <actualLength> bytes."
         ]
       },
+      "AES_IV_LENGTH" : {
+        "message" : [
+          "supports 16-byte CBC IVs and 12-byte GCM IVs, but got <actualLength> bytes for <mode>."
+        ]
+      },
+      "AES_SALTED_MAGIC" : {
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:1059]: Is it used somewhere? If not, please, remove it.

Comment (sweisdb) [reply core/src/main/resources/error/error-classes.json:1059]: I think this was re-added by mistake when I merged this on top of SPARK-43286. My mistake.

**Comment by MaxGekk** [comment]
@sweisdb Could you fix the code style issues:
```
Checkstyle checks failed at following occurrences:
Error:  src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionImplUtils.java:[56] (sizes) LineLength: Line is longer than 100 characters (found 114).
Error:  src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionImplUtils.java:[117] (sizes) LineLength: Line is longer than 100 characters (found 121).
Error:  src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionImplUtils.java:[118] (sizes) LineLength: Line is longer than 100 characters (found 102).
Error:  src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionImplUtils.java:[121] (sizes) LineLength: Line is longer than 100 characters (found 110).
Error:  src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionImplUtils.java:[122] (sizes) LineLength: Line is longer than 100 characters (found 104).
```

**Comment by MaxGekk** [comment]
> This change adds support for optional IV and AAD fields to aes_encrypt and aes_decrypt

@sweisdb Looking at the constructors of the `AesEncrypt` and `AesDecrypt` expressions, they still don't support new parameters. Are you going to update the expressions, correct?

**Comment by sweisdb** [comment]
@MaxGekk I am planning to doing the user-facing SQL expression changes in a followup to make each change more simple. I want to land this first.

**Comment by sweisdb** [comment]
There is an unrelated "sql - other tests" failing. I'm going to rebase and start everything over again.

**Comment by MaxGekk** [comment]
> I want to land this first.

ok. Let's modify PR's title and its description according to your actual changes.

**Review by MaxGekk — CHANGES_REQUESTED** [review]
LGTM in general. Please, adjust PR's description.

File: core/src/main/resources/error/error-classes.json [hunk 4]
```diff
@@ -5627,4 +5642,4 @@
       "Failed to get block <blockId>, which is not a shuffle block"
     ]
   }
-}
+}
```
[the remark this answers is not in the document]
Comment (sweisdb) [reply core/src/main/resources/error/error-classes.json:5645]: This class is generated by `SPARK_GENERATE_GOLDEN_FILES=1 build/sbt  "core/testOnly *SparkThrowableSuite -- -t \"Error classes are correctly formatted\""` to pass one of the unit tests which was checking that the output was properly formatted. 

I can manually revert this to add a newline at the end of the file, but if it causes the test to fail, I'll need to go back.

**Comment by sweisdb** [comment]
I updated the description to clarify that this change is just for `ExpressionImplUtilsSuite` and doesn't expose the user-facing changes in `aes_encrypt` and `aes_decrypt` yet.

**Comment by MaxGekk** [comment]
> this change is just for ExpressionImplUtilsSuite 
> Adds AES IV and AAD support to ExpressionImplUtilsSuite

@sweisdb Please, replace ExpressionImplUtilsSuite by ExpressionImplUtils. *Suite is the suffix of test suites but you changed not only tests.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @sweisdb.

**Pull request closed without merging**